### DISCUSSION TITLE
Masking no longer seems to work with the Composite method in Image::Magick for Perl

### DISCUSSION THREAD

**Discussion by shunn** [ImageMagick/ImageMagick#3544] [Help]
I have a Perl script on my website that is supposed to composite a masked image onto a base image. It worked fine until recently when I upgraded my web hosting plan. Previously my server had ImageMagick 6 (or maybe earlier) installed. The new server has ImageMagick 7.

Here's a simplified version of the script, which shows the basic flow and which, as I said, used to work fine:

```
use Image::Magick;

# initialize image objects
my $base = Image::Magick->new;
my $overlay = Image::Magick->new;
my $mask = Image::Magick->new;

# Read source images from disk
$base->Read('base.jpg'); # base image
$overlay->Read('overlay.jpg'); # overlay image
$mask->Read('mask.jpg'); # mask image

# Composite overlay image onto base image using mask image as mask
$base->Composite(
        image => $overlay,
        mask => $mask,
);

# Write target image to disk
$base->Write('result.jpg');
```

On the old server, `result.jpg` would contain the base image with a crisply masked section of the overlay image composited onto it. On the new server, however, `result.jpg` ends up containing the overlay image only. It was composited onto the base image with no masking at all. (And before you ask, yes, I have verified that the base image is being read from the disk correctly.)

My question is, is there something different I need to do to make the masking work with Composite in ImageMagick 7? Or is there perhaps something wrong with the updated Perl interface? Is this a known issue?

(I should add that I can use all the `compose` options for Composite successfully. When I use `compose` I get a result image that's a true composite, with the caveat that any included masking still fails.)

**Comment by urban-warrior** [comment]
See https://imagemagick.org/script/porting.php.  The page enumerates the differences between the two versions.  Try negating the mask ($mask->Negate()).  Does that work?

**Reply by shunn** [reply]
Thanks, but I just tried that at your suggestion and I'm afraid it didn't work.

**Comment by salamichel** [comment]
Hi @shunn i obsverved the same issue using http://www.fmwconcepts.com/imagemagick/tshirt/index.php

it works perfectly right on Version: `ImageMagick 6.9.10-68 Q16 x86_64 2021-02-24 https://imagemagick.org`

but now with Version: `ImageMagick 7.0.8-68 Q16 x86_64 2019-12-18 https://imagemagick.org`

I don't know how i can downgrade to 6.9x (on docker php:7.4-fpm-alpine)

**Reply by shunn** [reply]
Sounds like this may be a larger issue than just the Perl interface!
